### PR TITLE
GEOMESA-1188 Configurable interval for z3 dates

### PR DESCRIPTION
**Pull request by elahrvivaz**
Note: stacks on top of #971, please consider only last commit(s).

Signed-off-by: Emilio Lahr-Vivaz <elahrvivaz@ccri.com>

### CODE REVIEW

**Comment by elahrvivaz** [comment]
needs rebase

File: geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/stats/StatSerializer.scala [hunk 1]
```diff
@@ -273,15 +281,16 @@ object KryoStatSerializer {
...
     val dtgIndex  = input.readInt(true)
+    val period = if (version > 1) TimePeriod.withName(input.readString()) else TimePeriod.Week
     val length = input.readInt(true)
 
     val stat = if (immutable) {
-      new Z3Histogram(geomIndex, dtgIndex, length) with ImmutableStat
+      new Z3Histogram(geomIndex, dtgIndex, period, length) with ImmutableStat
     } else {
-      new Z3Histogram(geomIndex, dtgIndex, length)
+      new Z3Histogram(geomIndex, dtgIndex, period, length)
```
Comment (jnh5y) [inline geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/stats/StatSerializer.scala]: Ok, just for my sanity, did the various v1 stats have a time bound implied?  And now we are just being explicit about it?

File: geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/stats/Z3Frequency.scala [hunk 2]
```diff
@@ -12,39 +12,42 @@ import java.util.Date
...
   override type S = Z3Frequency
 
   private val mask = Frequency.getMask(precision)
+  private val sfc = Z3SFC(period)
+  private val timeToBin = BinnedTime.timeToBinnedTime(period)
 
   private [stats] val sketches = scala.collection.mutable.Map.empty[Short, CountMinSketch]
   private [stats] def newSketch: CountMinSketch = new CountMinSketch(eps, confidence, Frequency.Seed)
 
+
```
Comment (jnh5y) [inline geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/stats/Z3Frequency.scala]: Spare whitespace::)

File: geomesa-z3/src/main/scala/org/locationtech/geomesa/curve/BinnedTime.scala [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+  private def toDayAndMillis(date: DateTime): BinnedTime = {
+    val days = Days.daysBetween(Epoch, date)
+    val millisInDay = date.getMillis - Epoch.plusDays(days.getDays).getMillis
+    BinnedTime(days.getDays.toShort, millisInDay)
+  }
+
+  private def fromDayAndMillis(date: BinnedTime): DateTime = Epoch.plusDays(date.bin).plus(date.offset)
+
+  private def toWeekAndSeconds(time: Long): BinnedTime =
+    toWeekAndSeconds(new DateTime(time, DateTimeZone.UTC))
```
Comment (jnh5y) [inline geomesa-z3/src/main/scala/org/locationtech/geomesa/curve/BinnedTime.scala]: Joda object creation isn't free; any chance that it'd make sense to do this arithmetic manually?

Comment (elahrvivaz) [inline geomesa-z3/src/main/scala/org/locationtech/geomesa/curve/BinnedTime.scala]: we need it for the `Weeks.weeksBetween` and related method - there might be another way to do it without joda time but I wanted to keep things compatible with the current z3 - using just dates or longs might miss some of the leap days, etc.

File: geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/stats/Z3Frequency.scala (hunk 2 above)
Comment (elahrvivaz) [inline geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/stats/Z3Frequency.scala]: fixed!

File: geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/stats/StatSerializer.scala (hunk 1 above)
Comment (elahrvivaz) [inline geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/stats/StatSerializer.scala]: there's not a time bound, but the Z3-related stats partition things by time period. Internally we have a stat object (like a count min sketch or whatever) for each e.g. week, which results in less error when estimating counts. When we store the stat, we store each time period separately. Previously it was always weeks, but now it's tied to the Z3 interval.

**Comment by jnh5y** [comment]
Looks good to me once the other PR goes through.

**Comment by jahhulbert-ccri** [comment]
@elahrvivaz can you check the rebase and make sure this is still clean for merge.

**Comment by elahrvivaz** [comment]
@jahhulbert-ccri rebased again, thanks

**Pull request merged**